File: windows-app/src/ClipboardSync.App/Models/ProtocolModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using ClipboardSync.App.Util;

namespace ClipboardSync.App.Models;

public enum ContentType
{
    TEXT,
    URL,
    IMAGE,
    MIXED_UNSUPPORTED
}

public enum TransferState
{
    QUEUED,
    SENDING,
    AWAITING_ACK,
    ACKED,
    FAILED,
    DEFERRED
}

public sealed record ImageMetadata(
    int Width,
    int Height,
    long ByteSize,
    string ChecksumSha256,
    string Encoding,
    string? TransferId = null
);

public sealed record ClipboardEvent(
    string EventId,
    string SourceDeviceId,
    string OriginatedAtUtc,
    ContentType ContentType,
    string MimeType,
    long PayloadSizeBytes,
    string ContentHashSha256,
    string DedupeKey,
    TransferState TransferState,
    string? TextPayload = null,
    ImageMetadata? Image = null
);

public sealed record TransferDescriptor(
    string TransferId,
    string EventId,
    int TotalChunks,
    long TotalBytes,
    string ChecksumSha256
);

public sealed record TransferChunk(
    string TransferId,
    int ChunkIndex,
    string Base64Payload
);

public sealed record ProtocolEnvelope(
    string Type,
    string TimestampUtc,
    string? SessionId = null,
    string? DeviceId = null,
    string? Challenge = null,
    string? Response = null,
    string? Status = null,
    string? Reason = null,
    ClipboardEvent? Event = null,
    TransferDescriptor? Transfer = null,
    TransferChunk? Chunk = null
);

public sealed record PairingPayload(
    string DeviceId,
    string DisplayName,
    string ServiceName,
    string Host,
    int Port,
    string PairingCode,
    string CertificateSha256
);

public sealed record DiscoveryMessage(
    string Type,
    string? DeviceId = null,
    string? DisplayName = null,
    string? ServiceName = null,
    string? Host = null,
    int? Port = null,
    string? CertificateSha256 = null)
{
    public const string DiscoverType = "clipboard_sync_discover";
    public const string ResponseType = "clipboard_sync_device";
}

public sealed record NormalizedClipboardItem(
    ClipboardEvent Event,
    byte[]? ImageBytes,
    string PreviewText,
    string? PreviewUri = null,
    bool FromRemote = false
);

public sealed class RecentClipboardItem
{
    public required string EventId { get; init; }
    public required ContentType ContentType { get; init; }
    public required string PreviewText { get; init; }
    public string? PreviewUri { get; init; }
    public required long PayloadSizeBytes { get; init; }
    public required string SyncedAtUtc { get; init; }
    public required string DirectionLabel { get; init; }
    public required TransferState TransferState { get; set; }
    public required string Status { get; set; }
}

public sealed class SavedDeviceItem : ObservableObject
{
    private bool _selected;
    private bool _available;
    private bool _connected;

    public required string DeviceId { get; init; }
    public required string DisplayName { get; set; }
    public required string LastSeenUtc { get; set; }

    public bool Selected
    {
        get => _selected;
        set => SetProperty(ref _selected, value);
    }

    public bool Available
    {
        get => _available;
        set => SetProperty(ref _available, value);
    }

    public bool Connected
    {
        get => _connected;
        set => SetProperty(ref _connected, value);
    }

    public string Status => Connected ? "Connected" : Available ? "Available" : Selected ? "Selected" : "Saved";
}

public static class ProtocolJson
{
    public static JsonSerializerOptions Options { get; } = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = false,
        Converters =
        {
            new JsonStringEnumConverter()
        }
    };
}

File: ios-app/ClipboardSync/Core/ProtocolModels.swift
import Foundation

enum ContentType: String, Codable, CaseIterable {
    case text = "TEXT"
    case url = "URL"
    case image = "IMAGE"
    case mixedUnsupported = "MIXED_UNSUPPORTED"
}

enum TransferState: String, Codable, CaseIterable {
    case queued = "QUEUED"
    case sending = "SENDING"
    case awaitingAck = "AWAITING_ACK"
    case acked = "ACKED"
    case failed = "FAILED"
    case deferred = "DEFERRED"
}

enum TransportKind: String, Codable, CaseIterable {
    case none = "NONE"
    case lan = "LAN"
    case bleFallback = "BLE_FALLBACK"

    var label: String {
        switch self {
        case .lan: return "Wi-Fi / LAN"
        case .bleFallback: return "Bluetooth fallback"
        case .none: return "No active transport"
        }
    }
}

struct ImageMetadata: Codable, Equatable {
    var width: Int
    var height: Int
    var byteSize: Int64
    var checksumSha256: String
    var encoding: String
    var transferId: String?
}

struct ClipboardEvent: Codable, Equatable, Identifiable {
    var id: String { eventId }

    var eventId: String
    var sourceDeviceId: String
    var originatedAtUtc: String
    var contentType: ContentType
    var mimeType: String
    var payloadSizeBytes: Int64
    var contentHashSha256: String
    var dedupeKey: String
    var transferState: TransferState
    var textPayload: String?
    var image: ImageMetadata?
}

struct TransferDescriptor: Codable, Equatable {
    var transferId: String
    var eventId: String
    var totalChunks: Int
    var totalBytes: Int64
    var checksumSha256: String
}

struct TransferChunk: Codable, Equatable {
    var transferId: String
    var chunkIndex: Int
    var base64Payload: String
}

struct ProtocolEnvelope: Codable, Equatable {
    var type: String
    var timestampUtc: String
    var sessionId: String?
    var deviceId: String?
    var challenge: String?
    var response: String?
    var status: String?
    var reason: String?
    var event: ClipboardEvent?
    var transfer: TransferDescriptor?
    var chunk: TransferChunk?

    init(
        type: String,
        timestampUtc: String = Date.utcNowString,
        sessionId: String? = nil,
        deviceId: String? = nil,
        challenge: String? = nil,
        response: String? = nil,
        status: String? = nil,
        reason: String? = nil,
        event: ClipboardEvent? = nil,
        transfer: TransferDescriptor? = nil,
        chunk: TransferChunk? = nil
    ) {
        self.type = type
        self.timestampUtc = timestampUtc
        self.sessionId = sessionId
        self.deviceId = deviceId
        self.challenge = challenge
        self.response = response
        self.status = status
        self.reason = reason
        self.event = event
        self.transfer = transfer
        self.chunk = chunk
    }
}

struct PairingPayload: Codable, Equatable, Identifiable {
    var id: String { deviceId }

    var deviceId: String
    var displayName: String
    var serviceName: String
    var host: String
    var port: Int
    var pairingCode: String
    var certificateSha256: String
}

struct DiscoveryMessage: Codable, Equatable {
    static let discoverType = "clipboard_sync_discover"
    static let responseType = "clipboard_sync_device"

    var type: String
    var deviceId: String?
    var displayName: String?
    var serviceName: String?
    var host: String?
    var port: Int?
    var certificateSha256: String?
}

struct ResolvedPeerEndpoint: Codable, Equatable {
    var serviceName: String
    var host: String
    var port: Int
}

struct NormalizedClipboard {
    var event: ClipboardEvent
    var imageBytes: Data?
    var previewText: String
    var previewUri: String?
    var fromRemote: Bool
}

enum ProtocolJSON {
    static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        return encoder
    }()

    static let decoder: JSONDecoder = {
        JSONDecoder()
    }()
}

extension Date {
    static var utcNowString: String {
        ISO8601DateFormatter.clipboardSync.string(from: Date())
    }
}

extension ISO8601DateFormatter {
    static let clipboardSync: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter
    }()
}

